### ISSUE TITLE
Add flash attention 2

### ISSUE THREAD

**Issue by pphuc25** [PKU-YuanGroup/LanguageBind#19]
As explore the code, and in my knowledge (please correct if there are something wrong), the current code do not have flash attention in training but instead that the vanilla attention
I think flash attention is a low hanging fruit when training and eval will be faster but still the same result
Do you have any plan to apply flash attention to your code?

**Comment by LinB203** [comment]
Great idea! It is worth to try. Maybe we will release it with the **HUGE** version.

**Comment by LinB203** [comment]
Initially, I intended to use Xformers to accelerate training because Xformers support more GPUs compared to Flash Attention while maintaining similar efficiency. However, Xformers require tensor shapes that are multiples of 8. Since the text length in CLIP is 77, it does not meet this requirement. Therefore, I will proceed to try Flash Attention.

**Comment by LinB203** [comment]
https://github.com/Dao-AILab/flash-attention/issues/409

**Comment by LinB203** [comment]
According to this [issue](https://github.com/Dao-AILab/flash-attention/issues/409), Flash Attention currently does not support custom attention masks. However, OpenCLIP's text requires irregular attention masks. Additionally, the OpenCLIP team seems to be exploring the integration of Flash Attention [here](https://github.com/mlfoundations/open_clip/issues/317). Perhaps we can wait for a while to see if any developments arise.

**Comment by pphuc25** [comment]
cool information, thank you for provided

**Comment by pphuc25** [comment]
however, do you think the effort in extend the context length is worth?, when multiples of 8 there're so much advantages such as more suitable with TPU, mod 2 == 0 @LinB203

**Comment by LinB203** [comment]
In fact, the token length of 77 is not something I have set, but rather a setting of the CLIP model. I have not changed it because I do not want to modify any pretrained weights. I am not sure if changing it to a multiple of 8 would have an impact on performance. Additionally, to reduce training costs, I have tried incorporating `grad_checkpoint`, which has reduced the memory consumption from **31g to 11g** on V100 under the same batch size. The code will be released along with the **HUGE** version.

> however, do you think the effort in extend the context length is worth?, when multiples of 8 there're so much advantages such as more suitable with TPU, mod 2 == 0 @LinB203

**Comment by LinB203** [comment]
Larger models can now be trained by turning on [`--grad-checkpointing`](https://github.com/PKU-YuanGroup/LanguageBind/blob/main/scripts/video_language/train_1.5_huge.sh#L12).
